File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        List<Integer>[] bucket = new List[nums.length+1];
        Map<Integer, Integer> map= new HashMap<>();

        for(int i: nums){
            map.put(i, map.getOrDefault(i,0)+1);
        }
        for(int i: map.keySet()){
            int frequency = map.get(i);
            if(bucket[frequency]==null){
                bucket[frequency] = new ArrayList<>();
            }
            bucket[frequency].add(i);
        }

        int [] ans = new int[k];
        int counter =0;
        for(int i=bucket.length-1; i>=0 && counter<k; i--){
            if(bucket[i]!= null){
                for(int j: bucket[i] ){
                    ans[counter]= j;
                    counter ++;
                }
            }
        }

        return ans;
    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution(object):
    def topKFrequent(self, nums, k):
        counter = Counter(nums)
        mapper = [0]* (len(nums)+1)
        result = []
        for key, freq in counter.items():
            if mapper[freq]==0:
                mapper[freq] = [key]
            else:
                mapper[freq].append(key)
        
        for i in range(len(nums),-1,-1):
            if mapper[i]!=0:
                result.extend(mapper[i])
            if len(result)==k:
                break
        return result